pkg/mongodb/marshal.go

```diff
@@ -2,68 +2,258 @@ package mongodb
-		}
-		valueStr, ok := valueI.(string)
-		if !ok {
-			return
+var ErrPathNotFound = errors.New("path not found")
+
+// ConvertStringValueToInt64 converts string values to int64 in a JSON map based on provided paths.
+// It iterates over the specified paths in the JSON map and converts the string values found at those paths to int64 values.
+// If permitMissingPaths is set to true, missing paths in the JSON map will be ignored and the modified JSON map will be returned.
+// If permitMissingPaths is set to false, an error will be returned if any of the specified paths are not found in the JSON map.
+// The function returns the updated JSON map with the converted int64 values.
+// If an error occurs during the conversion, the partially modified JSON map is returned along with the error.
+func ConvertStringValueToInt64(jsonMap any, permitMissingPaths bool, paths ...string) (any, error) {
+	for _, path := range paths {
+		newMap, err := convertPath(jsonMap, permitMissingPaths, path)
+		if err != nil {
+			return jsonMap, err
 		}
-		value, err := strconv.ParseInt(valueStr, 10, 64)
+		jsonMap = newMap
+	}
+	return jsonMap, nil
+}
+
+func handleSlice(slice []any, permitMissingPaths bool, remainingParts []string) ([]any, error) {
+	var (
+		parents []any
+		errs    *multierror.Error
+	)
+
+	for _, item := range slice {
+		p, err := findParents(item, permitMissingPaths, remainingParts)
 		if err != nil {
-			return
+			errs = multierror.Append(errs, err)
+		} else if p != nil {
+			parents = append(parents, p...)
 		}
-		json[path] = value
-		return
 	}
 
-	elemPath := path[:pos]
-	elem, ok := json[elemPath]
-	if !ok {
-		return
-	}
-	elemArray, ok := elem.([]interface{})
-	if ok {
-		for i, elem := range elemArray {
-			elemMap, ok2 := elem.(map[string]interface{})
-			if !ok2 {
-				continue
+	if len(parents) == 0 {
+		return nil, errs.ErrorOrNil()
+	}
+
+	return parents, errs.ErrorOrNil()
+}
+
+func iterateOverMap(curr map[string]any, permitMissingPaths bool, part string) (any, error) {
+	var current any
+	if value, exists := curr[part]; exists {
+		current = value
+	} else if permitMissingPaths {
+		return nil, nil
+	} else {
+		return nil, fmt.Errorf("path segment %s: %w", part, ErrPathNotFound)
+	}
+	return current, nil
+}
+
+func iterateOverSlice(curr []any, permitMissingPaths bool, part string) (any, error) {
+	index, err := strconv.Atoi(part)
+	if err != nil {
+		return nil, fmt.Errorf("invalid array index %s", part)
+	}
+	if index < 0 || index >= len(curr) {
+		if permitMissingPaths {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("index out of range %d: %w", index, ErrPathNotFound)
+	}
+	return curr[index], nil
+}
+
+func findParents(current any, permitMissingPaths bool, parts []string) ([]any, error) {
+	for idx, part := range parts {
+		if part == "" {
+			continue
+		}
+		switch curr := current.(type) {
+		case map[string]any:
+			var err error
+			current, err = iterateOverMap(curr, permitMissingPaths, part)
+			if err != nil {
```

Comment:
```go
if err != nil {
  return nil, err
}
if current == nil {
  return nil, nil
}
```

These lines are the same for both cases, if we move them outside of the switch SC should stop complaining about function complexity

Reply:
fixed